### PR TITLE
[Quantization] Options to quantize input/output placeholders.

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
- Add options to quantize separately the input/output placeholders. Sometimes we don`t want to quantize all the placeholders (e.g. when the model ends with a Softmax which does not have a quantized implementation). Having separate options for inputs/ouputs adds extra flexibility.
- I needed to add in the **generateQuantizationInfos** the missing TQP for the output placeholders which do not have profiling information (the output nodes are not profiled). Without this the quantization of the output placeholders will not work because the TQP will not be found.
- Replaced in the bundle header generation the manual shape string derivation with **type.toString()** which already includes the shape and additionally for the quantized types the scales and offsets.

**Documentation**
Document parameters in AOT.md.

**Test Plan**
Unit test in QuantizationTest.

**Question**
Should I provide in the **quantizeFunction** also the **PlaceholderBindings** so that during the placeholder quantization, the backing tensors (if allocated) are also quantized?

### CODE REVIEW

File: lib/Quantization/Quantization.cpp [hunk 1]
```diff
@@ -737,6 +737,45 @@ class FunctionQuantizer : public FunctionConverter {
...
+        break;
+      }
+    }
+    return isOutputPlaceholder;
+  }
+
+  /// Function to quantize input placeholders. If the placeholder \p bindings
+  /// is given then also the backing tensors (if allocated) will be quantized.
+  void quantizeInputPlaceholders(PlaceholderBindings *bindings) {
+    for (auto ph : mod_.getPlaceholders()) {
```
Comment (jfix71) [inline lib/Quantization/Quantization.cpp]: nit: `auto *ph`

File: lib/Quantization/Quantization.cpp [hunk 2]
```diff
@@ -737,6 +737,45 @@ class FunctionQuantizer : public FunctionConverter {
...
+      }
+    }
+    cleanUp();
+    assert(function_.verify() && "Conversion led to invalid function");
+  }
+
+  /// Function to quantize output placeholders. If the placeholder \p bindings
+  /// is given then also the backing tensors (if allocated) will be quantized.
+  void quantizeOutputPlaceholders(PlaceholderBindings *bindings) {
+    for (auto ph : mod_.getPlaceholders()) {
```
Comment (jfix71) [inline lib/Quantization/Quantization.cpp]: nit: `auto *ph`

File: lib/Quantization/Quantization.cpp [hunk 3]
```diff
@@ -1037,6 +1092,16 @@ void quantizeFunction(Function *F, const QuantizationConfiguration &quantConfig,
   if (quantConfig.enableRowwise) {
     quantizer.enableRowwise();
   }
+
+  // Quantize input placeholders.
+  if (quantConfig.quantizeInputPlaceholders) {
+    quantizer.quantizeInputPlaceholders(nullptr);
```
Comment (jfix71) [inline lib/Quantization/Quantization.cpp]: nit: `/* bindings */ nullptr`

File: lib/Quantization/Quantization.cpp [hunk 4]
```diff
@@ -1037,6 +1092,16 @@ void quantizeFunction(Function *F, const QuantizationConfiguration &quantConfig,
...
+  }
+
+  // Quantize output placeholders.
+  if (quantConfig.quantizeOutputPlaceholders) {
+    quantizer.quantizeOutputPlaceholders(nullptr);
```
Comment (jfix71) [inline lib/Quantization/Quantization.cpp]: nit: `/* bindings */ nullptr`

**Comment by mciprian13** [comment]
@jfix Actually that is exactly what I needed (if the network ends in a Softmax then the output will not be converted with the options which is what I need). I replaced the code to use the already available graph optimization passes with a CLI opt in Loader.
Have a look and if fine by you, let's have this landed!
Thanks

**Comment by jfix71** [comment]
@mciprian13 I'm totally fine with reusing the optimization -- but I think this option is still going to cause issues when used without bundle mode right? E.g. if we merge a `QuantizeNode` into an input `Placeholder`, then we need to make sure the backing input Tensor is quantized in a pre-processing step if it was not already quantized. If this is not interesting/important to your use cases I'd at least like some sort of error printed/elegant exit when `!emittingBundle() && ! convertPlaceholdersOpt`.

**Comment by mciprian13** [comment]
@jfix I understand the problem but I thought the optimization pass takes care of the tensor conversion. I have a lot on my plate lately so I think I`ll go (for now) with option 2. I inserted the verification you mentioned.

**Pull request closed without merging**